### PR TITLE
Implement more sophisticated public path testing

### LINKED ISSUE #1467: Latest version does not respect pathbase
## Description
The version bundled with the latest release of ASF tries to load the bundle.js file from `/scripts/main.{hash}.bundle.js` instead of `/{pathBase}/scripts/main.{hash}.bundle.js`.

## Steps to reproduce
- Install latest version of ASF 
- Add custom IPC configuration with `PathBase` option
- Add reverse proxy in front of ASF, making sure that the root directory (in particular `/scripts`) will not be forwarded to ASF
- Access ASF-ui via reverse proxy

## Expected behavior
ASF-ui loads and displays correctly

## Current behavior
ASF-ui fails to load its main bundle due to it being loaded from the root directory.

## Additional information
[example_config_files.zip](https://github.com/JustArchiNET/ASF-ui/files/6758012/example_config_files.zip)

### PR DESCRIPTION
**Pull request by Aareksio**
## Description
TLDR: Respect public path

I considered the following possible options of running ASF-ui:
- under root directory, eg `http://localhost:1242`
- under sub-direcory, eg. `http://localhost:1242/ASF`
- under multi-level sub-directory, eg. `http://localhost:1242/path/to/ASF`

Previous public path method relied on `/page` in the URL to find the correct base path on first try, new method attempts to load the main script (`main.[hash].bundle.js`) from each level of the hierarchy.

Consider the complex URL: `http://localhost:1242/path/to/ASF/bots/test/config`
The public path is `/path/to/ASF`, it should open `/bots/:bot/config` page

The proposed solution will attempt to load the main script from the following paths:
- `/scripts/main.bundle.js` - default attempt - miss
- `/path/to/ASF/bots/test/config/scripts/main.bundle.js` - miss
- `/path/to/ASF/bots/test/scripts/main.bundle.js` - miss
- `/path/to/ASF/bots/scripts/main.bundle.js` - miss
- `/path/to/ASF/scripts/main.bundle.js` - hit

As you can see, it traverses down the path tree searching for the `scripts` directory. This allows us to drop the `/page` prefix, as well as be much more accurate when handling complex prefixes.

Furthermore, this proposal do not rely on `error` event alone to determine loading script failed. The browser does **not** emit `error` event when webserver responds with `200 OK`, even if the actual content of the request is not JavaScript, but, for example, HTML (catch-all index.html). We listen for `load` event, then check for __ASF_UI_LOADED__ property, which is set in index.js file. If not present, we treat `load` as `error` and continue the search.

## Additional information
Tests can be done using nginx reverse proxy with configuration provided by Abry in #1467, if too lazy to use nginx, you can launch the script in `scripts/reverse-proxy` and use `127.0.0.1:5000/ASF` in your browser.

Please share your feedback, before we ship another the breaking change to mainline ASF users :)

## Checklist

- [x] My pull request is not a duplicate
- [x] I added a descriptive title to this pull request
- [x] I added a concise description or a self-explanatory screenshot to this pull request
- [x] My code follows the code style of this project
- [x] I have performed a self-review of my own code

Closes #1467

### CODE REVIEW

**Comment by Abrynos** [comment]
Works for me. Thanks for the quick fix.

**Comment by JustArchi** [comment]
@Aareksio why the first attempt to try is the root `/` as opposed to current directory? What if user puts some crap there under the same name? You can't assume that if ASF is inside `/ASF` path then `/scripts/{whatever}.js` doesn't exist or exists with the content you want.

The logic itself is very nice but I'd argue that you should try to traverse the path from the current directory going down (`..`) until you find what you want. This way we're sure that whatever we access along the way belongs to us, until eventually we run out of the options and reach the root.

**Comment by Aareksio** [comment]
@JustArchi I was considering this, my thoughts are:

1. We should optimize for the most common use case, which is hosting at base path.
2. Putting `./` as default public path, may result in browser loading and executing the bundle before inline JS has time to fix the webpack's public path. Just as previously in #1467 they had the script _fix_ the path, but it was already too late. I did not want to risk it another bug.
3. None of the above really matter, because we now check for `__ASF_UI_LOADED__` property. Unless your script, which already miraculously contains the exact 7 random characters, sets this property, we ignore it. 

I see how this opens a possibility of XSS on shared hosts though. The script from `/scripts/` is loaded in the browser. Even if it is ignored by the path algorithm, it may contain XSS code. Nothing prevents the attacker to setup a logger for ASF-ui data and input.

**Comment by JustArchi** [comment]
> None of the above really matter, because we now check for __ASF_UI_LOADED__ property. Unless your script, which already miraculously contains the exact 7 random characters, sets this property, we ignore it.

Excellent, this should bullet-proof enough, as in addition to the file itself we actually check if it's the file we want. I don't believe we need to bullet-proof this more then, you can consider my point void then.

> I see how this opens a possibility of XSS on shared hosts though. The script from /scripts/ is loaded in the browser. Even if it is ignored by the path algorithm, it may contain XSS code. Nothing prevents the attacker to setup a logger for ASF-ui data and input.

We have the `hash` of the file we're loading, right? Is it possible to fetch the content of the file prior to "sourcing" (executing) it? If yes, we could calculate `md5` or whatever hashing algorithm webpack is using to determine if the file doesn't only have valid name, but also valid content for us. This way you bullet-proof against XSS, because if the file gets modified in any way, then ASF-ui will consider it to be "not our file" (same as it not existing), and move along without executing it.

**Comment by Aareksio** [comment]
> We have the hash of the file we're loading, right? Is it possible to fetch the content of the file prior to "sourcing" (executing) it? If yes, we could calculate md5 or whatever hashing algorithm webpack is using to determine if the file doesn't only have valid name, but also valid content for us. This way you bullet-proof against XSS, because if the file gets modified in any way, then ASF-ui will consider it to be "not our file" (same as it not existing), and move along without executing it.

We can not ignore the file from userland. If the browser loads it, we can not prevent execution from arbitrary JS code. There is such security built in the `<script>` tag: <https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity> I'll see what can be done.

**Comment by JustArchi** [comment]
This SRI looks like it achieves exactly what I have in mind, and webpack probably already has existing mechanisms to work with it without much problems, it sounds like a good idea to make use of, as we don't need unnecessary XSS in ASF-ui, especially if it can be solved this easily.
